### PR TITLE
parser: fix *TEXT BYTE | ASCII | UNICODE syntax

### PR DESCRIPTION
**Pull request by hey-kong**
#### Fix compatibility problem about *TEXT BYTE | ASCII | UNICODE syntax 

Issue: [pingcap/parser#529](https://github.com/pingcap/parser/issues/529)

#### MySQL Syntax:
```
type:
...
        | TINYTEXT_SYN opt_charset_with_opt_binary
          {
            $$= NEW_PTN PT_blob_type(Blob_type::TINY, $2.charset,
                                     $2.force_binary);
          }
        | TEXT_SYM opt_field_length opt_charset_with_opt_binary
          {
            $$= NEW_PTN PT_char_type(Char_type::TEXT, $2, $3.charset,
                                     $3.force_binary);
          }
        | MEDIUMTEXT_SYM opt_charset_with_opt_binary
          {
            $$= NEW_PTN PT_blob_type(Blob_type::MEDIUM, $2.charset,
                                     $2.force_binary);
          }
        | LONGTEXT_SYM opt_charset_with_opt_binary
          {
            $$= NEW_PTN PT_blob_type(Blob_type::LONG, $2.charset,
                                     $2.force_binary);
          }
...
        | LONG_SYM opt_charset_with_opt_binary
          {
            $$= NEW_PTN PT_blob_type(Blob_type::MEDIUM, $2.charset,
                                     $2.force_binary);
          }
```
```
opt_charset_with_opt_binary:
...
        | ascii
          {
            $$.charset= $1;
            $$.force_binary= false;
          }
        | unicode
          {
            $$.charset= $1;
            $$.force_binary= false;
          }
        | BYTE_SYM
          {
            $$.charset= &my_charset_bin;
            $$.force_binary= false;
          }
```
### Bad SQL Case:
```
create table t (a longtext unicode)
create table t (a mediumtext unicode)
create table t (a tinytext ascii)
create table t (a text byte)
create table t (a long byte, b text unicode)
create table t (a text unicode, b mediumtext ascii, c int)
create table t (a int, b text unicode, c mediumtext ascii)
```
### Check List

Tests
- Unit test

### CODE REVIEW

File: parser.y [hunk 1]
```diff
@@ -493,6 +493,7 @@ import (
 	truncate	"TRUNCATE"
 	unbounded	"UNBOUNDED"
 	uncommitted	"UNCOMMITTED"
+	unicodeType "UNICODE"
```
Comment (leoppro) [inline parser.y]:
```suggestion
	unicodeType	"UNICODE"
```

File: parser.y [hunk 2]
```diff
@@ -8965,6 +8966,32 @@ BlobType:
 		x := types.NewFieldType(mysql.TypeMediumBlob)
 		$$ = x
 	}
+|	"TINYTEXT" "BYTE"
```
Comment (leoppro) [inline parser.y]: add a rule : `ByteOpt`？

**Comment by leoppro** [comment]
resolve conflicts please

File: parser.y [hunk 3]
```diff
@@ -9111,6 +9138,20 @@ OptBinary:
 			Charset:  "",
 		}
 	}
+|   "ASCII"
```
Comment (tangenta) [inline parser.y]: Please indent these lines with tab.

File: parser.y (hunk 2 above)
Comment (tangenta) [reply parser.y]: These rules should be added to `TextType`.

File: parser.y (hunk 3 above)
Comment (tangenta) [reply parser.y]: Both `ASCII` and `UNICODE` should not belong to `OptBinary`.

File: misc.go [hunk 4]
```diff
@@ -613,6 +613,7 @@ var tokenMap = map[string]int{
 	"TYPE":                     tp,
 	"UNBOUNDED":                unbounded,
 	"UNCOMMITTED":              uncommitted,
+	"UNICODE":                  unicodeType,
```
Comment (leoppro) [inline misc.go]: should we?
```suggestion
	"UNICODE":                  unicode,
```

Comment (hey-kong) [reply misc.go]: After I do it, `make test` will fail:
```
...
./lexer.go:21: unicode redeclared in this block
        previous declaration at ./parser.go:635
...
./yy_parser.go:21: unicode redeclared as imported package name
        previous declaration at ./parser.go:635
...
```
"unicode" package can't be imported.

**Comment by tangenta** [comment]
@hey-kong Could you please add the following cases to the test?
```sql
create table t (a long ascii, b long unicode)
create table t (a long character set utf8mb4, b long charset utf8mb4, c long char set utf8mb4)
```

By the way, according to the [documentation](https://dev.mysql.com/doc/refman/8.0/en/blob.html),
> LONG and LONG VARCHAR map to the MEDIUMTEXT data type.

File: misc.go (hunk 4 above)
Comment (leoppro) [reply misc.go]: how about we name it unicodeSym

File: parser.y [hunk 5]
```diff
@@ -9037,6 +9038,32 @@ TextType:
...
+		$$ = &ast.OptBinary{
+			IsBinary: false,
+			Charset:  charset.CharsetLatin1,
+		}
+	}
+|	"UNICODE"
+	{
+		$$ = &ast.OptBinary{
+			IsBinary: false,
+			Charset:  charset.CharsetUTF8,
```
Comment (tangenta) [inline parser.y]: The following behaviors are specified in the [documentation](https://dev.mysql.com/doc/refman/8.0/en/string-type-overview.html) of MySQL:
> The ASCII attribute is shorthand for CHARACTER SET latin1.
The UNICODE attribute is shorthand for CHARACTER SET ucs2.

However, TiDB does not support charset `ucs2`. Should we ignore `UNICODE` in this case?
@kennytm @leoppro PTAL

Comment (kennytm) [reply parser.y]: IMO it should be successfully parsed but cause TiDB itself to error with 

```
ERROR 1115 (42000): Unknown character set: 'ucs2'
```

**Comment by leoppro** [comment]
@hey-kong  resolve conflicts please

**Comment by hey-kong** [comment]
> @hey-kong resolve conflicts please

ok

**Pull request merged**